newsroom/search/service.py

```diff
@@ -842,56 +843,60 @@ def prepend_embargoed_items_to_response(self, response, req, lookup):
-        aggs = {"topics": {"filters": {"filters": {}}}}
-        queried_topics = []
-
-        for topic in topics:
-            user = users.get(str(topic["user"]))
-            if not user:
-                continue
+        topic_matches = []
+        topics_checked = set()
 
+        for user in users.values():
+            aggs = {"topics": {"filters": {"filters": {}}}}
             company = companies.get(str(user.get("company", "")))
-
-            topic_query = self.get_topic_query(topic, user, company, query)
-            if not topic_query:
+            # there will be one base search for a user with aggs for user topics
+            search = self.get_topic_query(None, user, company, query=query)
+            if not search:
                 continue
+            queried_topics = []
+            for topic in topics:
+                user_id = str(topic["user"])
+                if not user_id or str(user["_id"]) != user_id:
+                    continue
+                if topic["_id"] in topics_checked:
+                    continue
+                topics_checked.add(topic["_id"])
+
+                topic_query = self.get_topic_query(topic, None, None)
+                if not topic_query:
+                    continue
+
+                aggs["topics"]["filters"]["filters"][str(topic["_id"])] = topic_query.query
+                queried_topics.append(topic)
+            if not queried_topics:
+                continue
+            source = {"query": search.query}
+            source["aggs"] = aggs if aggs["topics"]["filters"]["filters"] else {}
+            source["size"] = 0
 
-            aggs["topics"]["filters"]["filters"][str(topic["_id"])] = topic_query.query
-
-            queried_topics.append(topic)
-
-        source = {"query": query}
-        source["aggs"] = aggs if aggs["topics"]["filters"]["filters"] else {}
-        source["size"] = 0
-
-        req = ParsedRequest()
-        req.args = {"source": json.dumps(source)}
-        topic_matches = []
-
-        try:
-            search_results = self.internal_get(req, None)
-
-            for topic in queried_topics:
-                if search_results.hits["aggregations"]["topics"]["buckets"][str(topic["_id"])]["doc_count"] > 0:
-                    topic_matches.append(topic["_id"])
+            req = ParsedRequest()
+            req.args = {"source": json.dumps(source)}
 
-        except Exception as exc:
-            logger.error(
-                "Error in get_matching_topics for query: {}".format(json.dumps(source)),
-                exc,
-                exc_info=True,
-            )
-            raise
+            try:
+                search_results = self.internal_get(req, None)
+
+                for topic in queried_topics:
+                    if search_results.hits["aggregations"]["topics"]["buckets"][str(topic["_id"])]["doc_count"] > 0:
+                        topic_matches.append(topic["_id"])
```

Comment:
I was thinking, to help with Python memory consumption and performance etc, what if we `yield` the topic here and consume it in [send_topic_notification_emails](https://github.com/superdesk/newsroom-core/blob/develop/newsroom/push.py#L904)?

This way the topic is yielded and processed in the one step, as opposed to constructing the list of topics to later iterate & filter those same topics?

Reply:
makes sense, would do it rather separately as we might need to backport this also into 2.5 so would rather keep it small